### PR TITLE
Shopping_cart_2006 Project

### PR DESCRIPTION
**Pull request by nicorithner**

### CODE REVIEW

**Comment by timomitchel** [comment]
Nice job committing throughout!

File: lib/product.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class Product
+    attr_reader :category, :name, :unit_price, :quantity
+    def initialize(category, name, unit_price, quantity )
```
Comment (timomitchel) [inline lib/product.rb:3]: Nice job here, but there is white space between `quantity` and the closing parentheses `)` that should be removed.

File: lib/product.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    def is_hoarded?
+        @hoarded
+    end
+
+    def hoard
+        @hoarded = true
+    end
+
+end
```
Comment (timomitchel) [inline lib/product.rb:23]: This looks great. Exactly the implementation we wanted to see.

File: lib/shopping_cart.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
+class ShoppingCart
+    attr_reader :name, :capacity, :products
+    def initialize(name, capacity)
+        @name = name
+        @capacity = capacity[0..1].to_i
```
Comment (timomitchel) [inline lib/shopping_cart.rb:5]: Nice job getting this to run, but this solution wouldn't work if we passed in `"400items"`. You actually don't need the range `[0..1]` just the `.to_i`

File: lib/shopping_cart.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        @products = []
+    end
+    
+    def add_product(product)
+        products << product
+    end
+
+    def details
+      cart_details = {name: @name, capacity: @capacity}
```
Comment (timomitchel) [inline lib/shopping_cart.rb:14]: No need for the local variable `cart_details`

File: lib/shopping_cart.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    def total_number_of_products
+        @products.sum { |product| product.quantity}
```
Comment (timomitchel) [inline lib/shopping_cart.rb:18]: Perfect enum for the job here.

File: lib/shopping_cart.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    def is_full?
+        if total_number_of_products >= @capacity
```
Comment (timomitchel) [inline lib/shopping_cart.rb:22]: Close, but not quite. This needed to be:
`total_number_of_products > @capacity`
And the conditional can be completely removed.

File: lib/shopping_cart.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            true
+        else 
+            false
+        end
+    end
+
+    def products_by_category(category)
+        @products.find_all {|p| p.category == category}
```
Comment (timomitchel) [inline lib/shopping_cart.rb:30]: Nice job with `find_all`. `|p|` should be `|product|`

File: lib/shopping_cart.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    def percentage_occupied
+        ((total_number_of_products.to_f / @capacity) * 100).round(2)
+    end
+
+    def sorted_products_by_quantity
+        @products.sort_by do |product|
```
Comment (timomitchel) [inline lib/shopping_cart.rb:38]: Close, but this puts the products in ascending order, but the spec wanted them in descending order, so we needed to reverse this array after the `sort_by` ran: 
`@products.sort_by { |product| product.quantity}.reverse`

File: lib/shopping_cart.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            product.quantity
+        end
+    end
+
+    def product_breakdown
```
Comment (timomitchel) [inline lib/shopping_cart.rb:43]: On the right track, but not quite. Take a look at my solution PR to see one way to approach this.

File: test/product_test.rb [hunk 10]
```diff
@@ -10,35 +10,44 @@ def test_it_exists
...
   end
+
+  def test_it_can_hoard
+    # skip
+    product = Product.new(:paper, 'toilet paper', 3.70, '10')
+
+    assert_equal false,  product.is_hoarded?
+    product.hoard
+    assert_equal true, product.is_hoarded?
+  end
```
Comment (timomitchel) [inline test/product_test.rb:52]: Nice robust testing.

File: test/shopping_cart_test.rb [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    cart = ShoppingCart.new("King Soopers", "30items")
+    product1 = Product.new(:paper, 'toilet paper', 3.70, '10')
+    product2 = Product.new(:meat, 'chicken', 4.50, '2')
+    product3 = Product.new(:paper, 'tissue paper', 1.25, '1')
+
+    cart.add_product(product1)
+    cart.add_product(product2)
+    cart.add_product(product3)
+
+    assert_equal 13, cart.total_number_of_products
```
Comment (timomitchel) [inline test/shopping_cart_test.rb:44]: These two methods should have their own tests. We don't want to conflate methods in the same test.

File: test/shopping_cart_test.rb [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    assert_equal false, cart.is_full?
+
+    product4 = Product.new(:produce, 'apples', 0.99, '20')
+    cart.add_product(product4)
+    assert_equal true, cart.is_full?
```
Comment (timomitchel) [inline test/shopping_cart_test.rb:63]: Nice testing here.

File: test/shopping_cart_test.rb [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    cart.add_product(product1)
+    cart.add_product(product2)
+    cart.add_product(product3)
+
+    assert_equal 43.33, cart.percentage_occupied
+
+    product4 = Product.new(:produce, 'apples', 0.99, '20')
+    cart.add_product(product4)
+
+    assert_equal [product3, product2, product1, product4]  ,cart.sorted_products_by_quantity
```
Comment (timomitchel) [inline test/shopping_cart_test.rb:96]: Remember to follow the spec very closely. The object ids in the interaction pattern had this array reversed.

File: lib/product.rb (hunk 1 above)
Comment (nicorithner) [reply lib/product.rb:3]: Thank you. Removed

File: lib/shopping_cart.rb (hunk 3 above)
Comment (nicorithner) [reply lib/shopping_cart.rb:5]: Thank you for teaching me that. Corrected

File: lib/shopping_cart.rb (hunk 4 above)
Comment (nicorithner) [reply lib/shopping_cart.rb:14]: Removed

File: lib/shopping_cart.rb (hunk 6 above)
Comment (nicorithner) [reply lib/shopping_cart.rb:22]: I can see that now. I can't believe I didn't noticed that. Thank you for pointing this out. I need to be more detailed.

Comment (nicorithner) [reply lib/shopping_cart.rb:22]: I removed the conditional

File: lib/shopping_cart.rb (hunk 8 above)
Comment (nicorithner) [reply lib/shopping_cart.rb:38]: I changed it and changed the test. Another one where I didn't read carefully. I'll be more careful next time.

File: test/shopping_cart_test.rb (hunk 11 above)
Comment (nicorithner) [reply test/shopping_cart_test.rb:44]: corrected

**Pull request closed without merging**